File: ImasLiveDB/Models/Anniversary.swift
import Foundation
import GRDB

/// ブランド/アプリの記念日 (サービス開始・アプリ稼働・アニメ放映・劇場版公開等)。
/// date は YYYY-MM-DD (年あり)。毎年同じMM-DDで再点灯し、表示時に当該年からN周年を計算する。
struct Anniversary: Codable, FetchableRecord, PersistableRecord, Identifiable, Hashable, Sendable {
    static let databaseTableName = "anniversaries"

    var id: String
    var brandId: String
    /// 表示テキスト (例: "アーケード版稼働", "デレステ配信開始")。
    var label: String
    /// 起点日 'YYYY-MM-DD'。
    var date: String
    /// AnniversaryKind raw value。
    var kind: String
    var sortOrder: Int

    enum CodingKeys: String, CodingKey {
        case id, label, date, kind
        case brandId = "brand_id"
        case sortOrder = "sort_order"
    }

    /// 起点年 (date の YYYY 部分)。
    var startYear: Int? {
        Int(date.prefix(4))
    }

    /// 指定年における周年数 (起点年含めない: 2005-07-26 → 2026年で21周年)。
    /// 起点より前の年は nil。
    func anniversaryYears(in year: Int) -> Int? {
        guard let startYear else { return nil }
        let n = year - startYear
        return n >= 0 ? n : nil
    }
}

enum AnniversaryKind: String, CaseIterable {
    case serviceStart = "service_start"
    case appStart = "app_start"
    case animeStart = "anime_start"
    case movieRelease = "movie_release"
    case cdDebut = "cd_debut"

    /// 表示アイコン (SF Symbol)。
    var systemImage: String {
        switch self {
        case .serviceStart: return "sparkles"
        case .appStart: return "iphone"
        case .animeStart: return "tv"
        case .movieRelease: return "film"
        case .cdDebut: return "opticaldisc"
        }
    }
}

File: ImasLiveDB-Android/app/src/main/kotlin/com/fugaif/imaslivedb/data/model/Anniversary.kt
package com.fugaif.imaslivedb.data.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

/** アニバーサリーの種別。raw は DB 格納値。 */
enum class AnniversaryKind(val raw: String) {
    SERVICE_START("service_start"),   // アーケード/モバイルサービス開始
    APP_START("app_start"),           // スマートフォンアプリ配信開始
    ANIME_START("anime_start"),       // TVアニメ放映開始
    MOVIE_RELEASE("movie_release"),   // 劇場版公開
    CD_DEBUT("cd_debut");             // CDデビュー

    companion object {
        fun from(raw: String): AnniversaryKind =
            entries.firstOrNull { it.raw == raw } ?: SERVICE_START
    }
}

/** MIGRATION_4_5 が作る idx_anniversaries_brand / idx_anniversaries_date と indices を一致させること
 *  (ずれると Room の期待スキーマ検証で起動時クラッシュする)。 */
@Entity(
    tableName = "anniversaries",
    indices = [
        Index(value = ["brand_id"], name = "idx_anniversaries_brand"),
        Index(value = ["date"], name = "idx_anniversaries_date"),
    ]
)
data class Anniversary(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "brand_id")
    val brandId: String,

    @ColumnInfo(name = "label")
    val label: String,

    /** YYYY-MM-DD 形式。起点年はこの先頭4文字。 */
    @ColumnInfo(name = "date")
    val date: String,

    @ColumnInfo(name = "kind")
    val kind: String,

    @ColumnInfo(name = "sort_order", defaultValue = "0")
    val sortOrder: Int = 0
) {
    /**
     * 指定年における周年数。起点年 > year なら null (未来記念日)。
     * 例: date="2017-06-29", year=2026 → 9
     */
    fun anniversaryYears(year: Int): Int? {
        val startYear = date.take(4).toIntOrNull() ?: return null
        val years = year - startYear
        return if (years < 0) null else years
    }
}
